lib/oauthdb.js

```diff
@@ -0,0 +1,177 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+/* Operations on OAuth database state.
+ *
+ * Currently this is not actually talking to a database,
+ * it's making authenticated calls to the fxa-oauth-server API
+ * to interrogate and manipulate its state, essentially treating
+ * fxa-oauth-server as a kind of backend micro-service.
+ *
+ * We want to work towards merging the fxa-oauth-server code
+ * directly into the main fxa-auth-server process, at which point
+ * this abstraction will convert into more direct db access.
+ *
+ */
+
+const Joi = require('joi')
+const JWT = require('jsonwebtoken')
+
+const P = require('./promise')
+const createBackendServiceAPI = require('./backendService')
+const error = require('./error')
+const validators = require('./routes/validators')
+
+// The oauth-server's error numbers overlap and conflict
+// with the auth-server's, so we have to map them to new ones.
+function mapOAuthError(err) {
+  if (err instanceof error) {
+    return err
+  }
+  switch (err.errno) {
+    case 101:
+      return error.unknownClientId(err.clientId)
+    case 119:
+      return error.staleAuthAt(err.authAt)
+    default:
+      return error.unexpectedError()
+  }
+}
+
+module.exports = (log, config) => {
+
+  const OAuthAPI = createBackendServiceAPI(log, config, 'oauth', {
+
+    getClientInfo: {
+      path: '/v1/client/:clientId',
+      method: 'GET',
+      validate: {
+        params: {
+          clientId: validators.clientId.required()
+        },
+        response: {
+          id: validators.clientId.required(),
+          name: Joi.string().max(25).regex(validators.DISPLAY_SAFE_UNICODE).required(),
+          trusted: Joi.boolean().required(),
+          image_uri: Joi.string().optional().allow(''),
+          redirect_uri: Joi.string().required().allow('')
+        }
+      }
+    },
+
+    getScopedKeyData: {
+      path: '/v1/key-data',
+      method: 'POST',
+      validate: {
+        payload: {
+          client_id: validators.clientId.required(),
+          assertion: validators.assertion.required(),
+          scope: validators.scope.required()
+        },
+        response: Joi.object().pattern(/^/, [
```

Comment:
Since I'm not in joi all the time, I'm trying to decipher this validation. The [joi docs for pattern](https://github.com/hapijs/joi/blob/v14.3.0/API.md#objectpatternpattern-schema) aren't helping much either. Can you explain?

Reply:
This is supposed to mean "for any key maching the regex `/^/`, the corresponding value must match this schema", and since all keys match `/^/` it applies to every key.  From the joi docs, I think I can re-write this to be a little clearer:

```
Joi.object().pattern(Joi.any(), Joi.object({
  identifier: ...etc...
}))
```

Does that help?